### ISSUE TITLE
Dockerfile ADD comment parsing is wonky.

### ISSUE THREAD

**Issue by pnathan** [moby/moby#16456]
Hi, 

We have a dockerfile that looks somewhat like

```
...
USER ubuntu
ADD . /home/ubuntu/build  # get the build files into our build area.
...
```

Which fails out with an error message...

```
Step 2 : USER ubuntu
 ---> Running in 526c78b6f7e2
 ---> f8360c1e04f6
Removing intermediate container 526c78b6f7e2
Step 3 : ADD . /home/ubuntu/build  # get the build files into our build area.
home/ubuntu/build: no such file or directory
```

Which of course is unclear, since we've already ensured build/ exists there. Further, it lost the leading `/`.  However, if we modify it to:

```
-ADD . /home/ubuntu/build # get the build files into our build area.
+ # get the build files into our build area.
+ADD . /home/ubuntu/build
```

Everything works! 

:sob: 

Can we get a proper #-to-end-of-line comment parser added?

Version information:

```
 $ docker version
Client:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:08:05 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:08:05 UTC 2015
 OS/Arch:      linux/amd64

```

**Comment by thaJeztah** [comment]
I don't think this is something that will be implemented; the Dockerfile syntax is kept as simple as possible; comments should always start at the first position of a line.

The problem here is that we won't be able to support it for all commands, for example, ` RUN something # foo` is something that should be handled by the *shell* in the container, and Docker should not strip those comments

TL;DR I understand your thinking here, but I fear this is not something we'll be able to support

**Comment by pnathan** [comment]
I would strongly suggest that a hard look be done here on the design intent. UNIX tradition is that comments can go most anywhere; it's always extremely surprising when coming from one format into Dockerfiles and finding that this normative rule is violated. It's not simple to debug when the modality is otherwise. 

To respond to the cited Dockerfile command RUN (and it's cousins ENTRYPOINT and CMD).

I'm not persuaded that RUN foo # bar should execute `/bin/bash -c "foo #bar"`; this should have the same semantic content as `/bin/bash -c "foo"`.  What specifically do we (the Docker community) gain by not having # be "unanchored, until end of line" comments?

I note as a sampling of prior art here: YAML, conf, bash, Perl, awk, systemd units, TOML,  and INI. 

Also noted: if what is lacking here is a PR to address this, I believe that could be provided.

**Comment by memory** [comment]
I am wracking my brain trying to think of a scenario where stripping comments at the parse stage would result in unanticipated behavior when passing commands off to the shell and am coming up blank:

```
$ sh -c 'echo foo # echo bar; echo baz'
foo
```

So not that.

```
$ sh -c 'echo foo \
> # echo bar; echo baz'
foo
```

Also not that.

```
$ sh -c 'echo foo; # echo bar\
> echo baz'
foo
baz
```

And not that; am I missing a case?

**Comment by thaJeztah** [comment]
Perhaps `echo \#foo > something`, `echo '#foo'`;

Although these could possibly be solved, it's something that should be given a proper thought. Even though `echo foo` and `echo foo #bar` have the same semantic; it *is* a different instruction being sent to the shell. It's the builders (Docker's) task to send instructions to the shell unmodified; wether or not that produces the same result, is not something that the "builder" should know.

/cc @duglin for builder

**Comment by duglin** [comment]
From the docs ( http://docs.docker.com/reference/builder/ ):
```
Docker will treat lines that begin with # as a comment. A # marker anywhere else in the line will be treated as an argument. This allows statements like:

# Comment
RUN echo 'we are running some # of cool things'
```
I would love to fix things like this but as @thaJeztah said the Dockerfile is, kind of, shut down for now so making these changes right now just isn't in the cards.

At some point I'm hoping we can start over with a better formatted Dockerfile, one that's more consistent across the commands and does what people expect (like deal with comments anywhere), but right now its just a wish.

**Comment by pnathan** [comment]
@duglin So, what's the way forward here? Is there a road map on when the freeze is lifted? (i.e., is that a Docker 2.0 plan?)  

Also, it is a little confounding, since I don't think  a correctly implemented comment stripper would remove functionality and it should be backwards compatible... Like I said, if the problem is a lack of  working code, a pull request can be submitted here. :-)

**Comment by duglin** [comment]
@pnathan I don't know when the freeze will be lifted. The only thing I know for sure is that there are things underway to move the builder to the cli.  I'm hoping that eventually that will lead to a totally new builder (based on work that @jlhawn is doing) that will allow us to fix a lot of these types of issues.

But most of that is still under discussion.

As for a trying to get (which seems like a minor change made), I'll leave that up to @icecrime and @tiborvass.

**Comment by thaJeztah** [comment]
Let me close this ticket, because there's no plans to implement this currently, and the Dockerfile syntax is now maintained in https://github.com/moby/buildkit/
